### DISCUSSION TITLE
Accessing iTunes Music folder in iMac

### DISCUSSION THREAD

**Discussion by dunxd** [music-assistant/support#2336] [Q&A] [answered]
I have all my files stored on my Mac running Sonoma 14.5 through iTunes. I have Music Assistant as an Add-on for Home Assistant.

I set up a Standard user called MusicAssistant on my Mac.
From System Settings I went to General > Sharing > File Sharing > i
Then I added the folder where my iTunes files are stored, and added the MusicAssistant user to the access list with Read/Write permissions.

Now I try to add the folder as a Filesystem (remote share) Provider giving the fqdn and share name, and the credentials for the user.
I get the following error message:
```
Connection failed for the given details: SMB mount failed with error: mount error(13): Permission denied
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs) and kernel log messages (dmesg)
```
The error message in the add-on logs is the same.

On the HA Host log I see the following (hostname and timestamp redacted):
```
CIFS: Attempting to mount //hostname/Music
CIFS: Status code returned 0xc000006d STATUS_LOGON_FAILURE
CIFS: VFS: \\hostname Send error in SessSetup = -13
CIFS: VFS: cifs_mount failed w/return code = -13
```
I'm copying and pasting the password from the same place when I am creating the user and setting up the Provider, so no typos can happen.

In Mac System Settings > Network > Firewall > Options File Sharing (SMB) is set to Allow incoming connections.

What am I doing wrong?

Additional info that may be relevant.  My media files are on an external USB drive connected to the Mac.
I am not able to add that folder as Network Storage on the HA host.
I am able to add a different shared folder (share2) as Network Storage, but that folder is stored on the internal hard drive and doesn't have music files in it.
I am not able share2 as a Filesystem (remote share) to Music Assistant.  I get a different error:
```
Connection failed for the given details: SMB mount failed with error: mount: bad usage
Try 'mount --help' for more information.
```
The error is with the mounting - I don't think that issue is related to the contents of the share.

I worked on [an Add-On that was going to mount SMB shares](https://github.com/dunxd/HomeAssistantAddons/issues/10), and couldn't figure out how to do it (something to do with Docker's security model).  Then Home Assistant started supporting Network Shares which could then be mapped to Add-ons, so I gave up as that worked.  However, now I am suspecting there is some issue with sharing from an external USB drive...

**Comment by OzGav** [comment]
You can try using the IP address instead of the host name. However, MA is just using the OS mount command which you can see from the error message. I would say you are the right track and this is something that you have to resolve at your end. Maybe try a dedicated network forum.

**Comment by dunxd** [comment]
I think there is some issue accessing Mac SMB shares if the folder is on a USB drive.  From a Windows PC, I can get to the Mac in Windows Explorer, see all the shares listed and access all of them except any on a USB drive. 

I've tested this on USB drives formated with APFS and Mac OS Extended, and the same issue arises - the share is visible but can't be opened even using an admin account.  I'll share what I find out here, as I can't be the only person wanting to expose a large iTunes music library on an external drive from a Mac to Music Assistant.

**Comment by dunxd** [comment, marked as answer]
Found the fix for this.  I needed to grant `/usr/sbin/smbd` Full Disk Access rights.  After doing that I was able to add the share.

Here's how to do it for macos Sonoma:
* System Settings > Privacy & Security
* Scroll down to Full Disk Access and click the >
* Scroll down and click the +
* Enter password
* CMD + Shift + . to show hidden files and folders
* Navigate to `/usr/sbin` then select `smbd`
* Now you can add the share on your external USB drives to Music Assistant!

**Reply by Peter-HVD** [reply]
Doesn't work for me. I know the permissions are correct, and can connect directly using the Finder on another Mac with the same login. I've also tried setting up a dedicated user on the target Mac, to no avail, and have even added the above suggested smbd file directly to the Mac's Firewall. As I am able to log in on a different machine, I can't see how this can be a Mac permissions error.

**Reply by OzGav** [reply]
MA just uses the Linux mount command. There are additional options you can set in the advanced settings. Maybe Google for the problem generally as this likely isn’t something in the MA code.

**Reply by dunxd** [reply]
Try connecting to your share with a Windows device for a more valid test. Mac to Mac sharing doesn't have to use SMB/CIFS. 

After I made the changes above Windows could also access the share. Have you definitely enabled SMB on your Mac?

> Mac System Settings > Network > Firewall > Options File Sharing (SMB) is set to Allow incoming connections.

**Reply by Peter-HVD** [reply]
@dunxd 

> Try connecting to your share with a Windows device for a more valid test. Mac to Mac sharing doesn't have to use SMB/CIFS.

I don't have a Windows device. My testing of the login and permissions, whilst on the Mac, did not use the usual Finder sidebar method of selecting the server and logging in. Rather I used Connect to Server on the Go menu, although I will grant you that I have no idea of that makes the difference :-) But the point I'm making is that the error is "permission denied" and yet the login details are proven to be correct.

> Have you definitely enabled SMB on your Mac?

Yes, I have double & triple checked that! 

@OzGav I'm sorry, but that's just about the most unhelpful response imaginable! I am not the only person saying that they are having problems so surely that should warrant some investigation? And, yes, whilst network shares can be very complicated with many things outside of MA that could stop it working, there is a lot that could be checked to ensure that the mount command is not being malformed, or is missing a, possibly obscure, setting or option?

**Reply by OzGav** [reply]
As I said you have full control of the mount command options in the settings. The devs use Macs.

**Reply by dunxd** [reply]
I found it helpful to try adding the share as a mount in Home Assistant from `Settings > System > Storage > Add Network Storage`. If that doesn't work then you know it isn't specific to Music Assistant and the error messages are a bit easier to find in Home Assistant. If it does work, then that is also useful information.

I got the permission denied message too, but the problem was that the Mac couldn't authenticate access to the folder because _the `smbd`process_ didn't have permission to access the folder - the user permissions don't even come into it if that is the case. I have no doubt that if I had tried to access the folder from a Mac it would have been successful as the protocol would not have been an issue. I'm not saying your issue is exactly the same but maybe it will point you in the right direction. 

[apple.stackexchange.com](https://apple.stackexchange.com/questions/472836/how-to-share-a-folder-stored-on-an-external-usb-drive) was a useful resource once I found out (thanks to Windows) that the issue wasn't specifically Music Assistant or Home Assistant.

**Reply by Peter-HVD** [reply]
> I found it helpful to try adding the share as a mount in Home Assistant from `Settings > System > Storage > Add Network Storage`.

Thanks for that tip - I tried just that and it worked first time in Home Assistant. I then immediately tried to add it to MA and *it worked*! So that's odd, and possibly a coincidence, but maybe that's a fix for anyone else: add it to HA first.

**Reply by dunxd** [reply]
I had only meant that as a troubleshooting step, but very happy it got you up and running!
